pkg/cloud-init/cloud-init.go

```diff
@@ -110,6 +110,15 @@ func IsValidCloudInitData(cloudInitData *CloudInitData) bool {
 	return cloudInitData != nil && cloudInitData.UserData != "" && (cloudInitData.NoCloudMetaData != nil || cloudInitData.ConfigDriveMetaData != nil)
 }
 
+func cloudInitUUIDFromVMI(vmi *v1.VirtualMachineInstance) string {
```

Comment:
What about using `VMI.spec.firmware.uuid`
IIRC then this value is derived from the name by default, thus it does not change between restarts, and should even stay stable if VMI is managed bby a VM, yet it will be different if the VM is cloned.

To me using this field would make the solution more self contained.

Reply:
While it's a valid suggestion to use `VMI.spec.firmware.uuid` wouldn't it be slightly confusing for users if it was provided to cloud-init as the instance UUID? I'd much rather see the code below used to return either the VirtualMachine UUID or VirtualMachineInstance UUID as not to confuse users who might end up encountering it used as if it was the instance UUID.

Reply:
@lyarwood why do you think it's confusing? I actually think it's better to use that because as Fabian said it's self contained.

What we need for the instance-id is just a stable UUID and VMI.spec.firmware.uuid is derived from the VM/VMI name and is stable.

Reply:
Sorry, I missed Lee's comment - Lee, in KubeVirt's terms, the VMI firmware uuid is derived from the VMI name (by default), thus I'm seeing the firmware uuid to really be the instance id - because they are usually coupled. WDYT?